File: packaging/agent_evidence_vectors/runmetrics.py
"""A validator for the Run object of draft-arsentev-agent-run-metrics-00.

The draft is an individual Internet-Draft (Informational, 10 September 2026)
defining a JSON interchange format for resource accounting of language-model
agent runs. Its text is vendored beside the corpus that cites it and pinned by
digest; each rule here is bound to one sentence of that text by a requirement
identifier the corpus mints (``ARM-R-nnn``), and ``rejections(document)``
returns the identifiers a Run object is rejected under. Only the sentences a
validator can read off one document are carried: the HTTP binding, the
privacy and security guidance, and the Collector-side behaviours are not
properties of a Report and are not rows.

``corpora/runmetrics.go`` is the same statement in Go and the parity test diffs
the two.
"""

from __future__ import annotations

import re
from typing import Any

STATUSES = ("running", "completed", "failed", "aborted")
ENDED = ("completed", "failed", "aborted")
STEP_KINDS = ("model_invocation", "tool_call")
RUN_MEMBERS = (
    "version", "run_id", "parent_run_id", "root_run_id", "start", "end", "status",
    "termination_reason", "agent", "models", "step_count", "steps", "totals",
    "subtree_totals", "cost", "labels", "errors", "revision", "trace_id",
)
USAGE_MEMBERS = (
    "input_tokens", "output_tokens", "cache_read_tokens", "cache_write_tokens",
    "reasoning_tokens", "billable_input_tokens", "billable_output_tokens", "cache_writes",
)
IDENTIFIERS = ("run_id", "parent_run_id", "root_run_id")
TIMESTAMP = re.compile(r"^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(\.[0-9]+)?Z$")
AMOUNT = re.compile(r"^-?[0-9]+(\.[0-9]+)?$")

R = {n: f"ARM-R-{n:03d}" for n in range(1, 23)}


def _is_int(value: Any) -> bool:
    return isinstance(value, int) and not isinstance(value, bool)


def _is_str(value: Any) -> bool:
    return isinstance(value, str)


def _is_obj(value: Any) -> bool:
    return isinstance(value, dict)


def _get(obj: dict[str, Any], key: str) -> Any:
    return obj.get(key)


def shape_errors(document: Any) -> list[str]:
    """Every way the document fails to be a Run object at all."""
    if not _is_obj(document):
        return ["the document is not a JSON object"]
    out: list[str] = []
    for member in ("version", "run_id", "start", "status", "step_count", "totals"):
        if member not in document:
            out.append(f"the Run carries no {member} member")
    if "steps" in document and not isinstance(document["steps"], list):
        out.append("steps is present and is not an array")
    for i, step in enumerate(document.get("steps") or []):
        if not _is_obj(step):
            out.append(f"steps[{i}] is not an object")
        elif not _is_int(step.get("index")) or not _is_str(step.get("kind")):
            out.append(f"steps[{i}] carries no integer index or no string kind")
    return out


def _rows_run_status(run: dict[str, Any], out: set[str]) -> None:
    if run["version"] != "1":
        out.add(R[1])
    status = run["status"]
    if status not in STATUSES:
        out.add(R[4])
    elif (status in ENDED) != ("end" in run):
        out.add(R[2])
    start, end = _get(run, "start"), _get(run, "end")
    if _is_str(start) and _is_str(end) and TIMESTAMP.match(start) and TIMESTAMP.match(end):
        if end < start:
            out.add(R[3])
    steps: Any = run.get("steps")
    if isinstance(steps, list) and _is_int(run["step_count"]) and run["step_count"] < len(steps):
        out.add(R[5])


def _rows_run_members(run: dict[str, Any], out: set[str]) -> None:
    if "root_run_id" in run and "parent_run_id" not in run and run["root_run_id"] != run["run_id"]:
        out.add(R[16])
    if any(m not in RUN_MEMBERS and not m.startswith("x-") for m in run):
        out.add(R[21])
    labels = _get(run, "labels")
    if labels is not None and (not _is_obj(labels) or not all(_is_str(v) for v in labels.values())):
        out.add(R[18])
    cost = _get(run, "cost")
    if _is_obj(cost) and not (_is_str(cost.get("amount")) and AMOUNT.match(cost["amount"])):
        out.add(R[17])


def _rows_timestamps_and_identifiers(run: dict[str, Any], out: set[str]) -> None:
    stamps = [run.get("start"), run.get("end")]
    for step in run.get("steps") or []:
        stamps.append(step.get("start"))
        stamps.append(step.get("end"))
    if any(s is not None and not (_is_str(s) and TIMESTAMP.match(s)) for s in stamps):
        out.add(R[19])
    names = [run.get(k) for k in IDENTIFIERS if k in run]
    for step in run.get("steps") or []:
        if "invocation_id" in step:
            names.append(step["invocation_id"])
        if _is_obj(step.get("tool")) and "name" in step["tool"]:
            names.append(step["tool"]["name"])
    if any(not (_is_str(n) and 0 < len(str(n)) <= 128) for n in names):
        out.add(R[20])


def _rows_usage(usage: Any, out: set[str]) -> None:
    if not _is_obj(usage):
        return
    counters = {k: v for k, v in usage.items() if k != "cache_writes"}
    if any(not _is_int(v) or v < 0 for v in counters.values()):
        out.add(R[9])
        return
    read = counters.get("cache_read_tokens", 0)
    write = counters.get("cache_write_tokens", 0)
    if "input_tokens" in counters:
        if read > counters["input_tokens"]:
            out.add(R[10])
        elif read + write > counters["input_tokens"]:
            out.add(R[11])
    reasoning = counters.get("reasoning_tokens", 0)
    if "output_tokens" in counters and reasoning > counters["output_tokens"]:
        out.add(R[22])
    writes = usage.get("cache_writes")
    if not isinstance(writes, list):
        return
    lifetimes = [w.get("lifetime") for w in writes if _is_obj(w)]
    if len(set(lifetimes)) != len(lifetimes):
        out.add(R[13])
    tokens = [w.get("tokens") for w in writes if _is_obj(w)]
    if all(_is_int(t) for t in tokens) and sum(tokens) != write:
        out.add(R[14])


def _rows_steps(run: dict[str, Any], out: set[str]) -> None:
    steps = run.get("steps") or []
    indexes = [s["index"] for s in steps]
    starts = [s.get("start") for s in steps]
    pairs = list(zip(starts, starts[1:], strict=False))
    ordered = all(_is_str(a) and _is_str(b) and a <= b for a, b in pairs)
    if len(set(indexes)) != len(indexes) or indexes != sorted(indexes) or not ordered:
        out.add(R[6])
    invocations: list[str] = []
    for step in steps:
        kind = step["kind"]
        if kind == "model_invocation":
            if not (_is_obj(step.get("usage")) and _is_obj(step.get("model"))):
                out.add(R[7])
            if _is_str(step.get("invocation_id")):
                invocations.append(step["invocation_id"])
        elif kind == "tool_call" and (not _is_obj(step.get("tool")) or "usage" in step):
            out.add(R[8])
        _rows_usage(step.get("usage"), out)
    if len(set(invocations)) != len(invocations):
        out.add(R[15])


def _rows_totals(run: dict[str, Any], out: set[str]) -> None:
    totals = run["totals"]
    _rows_usage(totals, out)
    steps = run.get("steps")
    if not isinstance(steps, list) or not _is_obj(totals) or run["step_count"] != len(steps):
        return
    usages = [s["usage"] for s in steps if _is_obj(s.get("usage"))]
    if any(not _is_int(v) for u in usages for k, v in u.items() if k != "cache_writes"):
        return
    members = {k for u in usages for k in u if k != "cache_writes"} | {
        k for k in totals if k != "cache_writes"
    }
    for member in members:
        expected = sum(u.get(member, 0) for u in usages)
        if totals.get(member) != expected:
            out.add(R[12])
            return


def rejections(document: dict[str, Any]) -> list[str]:
    """The requirement identifiers this Run object is rejected under, sorted."""
    out: set[str] = set()
    _rows_run_status(document, out)
    _rows_run_members(document, out)
    _rows_timestamps_and_identifiers(document, out)
    _rows_steps(document, out)
    _rows_totals(document, out)
    return sorted(out)

File: corpora/runmetrics.go
package corpora

import (
	"encoding/json"
	"regexp"
	"sort"
	"strings"
)

// armRejections judges a Run object of draft-arsentev-agent-run-metrics-00.
// It is the Go statement of packaging/agent_evidence_vectors/runmetrics.py:
// the same sentences, bound to the same ARM-R-nnn identifiers, in the same
// order, so the parity test can diff the two over the corpus.

var (
	armStatuses   = set("running", "completed", "failed", "aborted")
	armEnded      = set("completed", "failed", "aborted")
	armRunMembers = set("version", "run_id", "parent_run_id", "root_run_id", "start", "end", "status",
		"termination_reason", "agent", "models", "step_count", "steps", "totals",
		"subtree_totals", "cost", "labels", "errors", "revision", "trace_id")
	armIdentifiers = []string{"run_id", "parent_run_id", "root_run_id"}
	armTimestamp   = regexp.MustCompile(`^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(\.[0-9]+)?Z$`)
	armAmount      = regexp.MustCompile(`^-?[0-9]+(\.[0-9]+)?$`)
)

func armR(n int) string { return "ARM-R-" + pad3(n) }

func pad3(n int) string {
	s := "000" + itoa(n)
	return s[len(s)-3:]
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	digits := ""
	for n > 0 {
		digits = string(rune('0'+n%10)) + digits
		n /= 10
	}
	return digits
}

type armRejects map[string]bool

func (r armRejects) add(n int) { r[armR(n)] = true }

func armShapeErrors(value any) []string {
	document, ok := value.(map[string]any)
	if !ok {
		return []string{"the document is not a JSON object"}
	}
	var out []string
	for _, member := range []string{"version", "run_id", "start", "status", "step_count", "totals"} {
		if _, present := document[member]; !present {
			out = append(out, "the Run carries no "+member+" member")
		}
	}
	if raw, present := document["steps"]; present {
		steps, ok := raw.([]any)
		if !ok {
			out = append(out, "steps is present and is not an array")
		}
		for i, step := range steps {
			object, ok := step.(map[string]any)
			if !ok {
				out = append(out, "steps["+itoa(i)+"] is not an object")
				continue
			}
			_, indexOK := intValue(object["index"])
			if !indexOK || !isStr(object["kind"]) {
				out = append(out, "steps["+itoa(i)+"] carries no integer index or no string kind")
			}
		}
	}
	return out
}

func armSteps(run map[string]any) []map[string]any {
	raw, _ := run["steps"].([]any)
	out := make([]map[string]any, 0, len(raw))
	for _, step := range raw {
		out = append(out, step.(map[string]any))
	}
	return out
}

func armTimestampOK(v any) bool {
	s, ok := v.(string)
	return ok && armTimestamp.MatchString(s)
}

func armRowsRunStatus(run map[string]any, out armRejects) {
	if version, _ := run["version"].(string); version != "1" {
		out.add(1)
	}
	status, _ := run["status"].(string)
	_, hasEnd := run["end"]
	switch {
	case !armStatuses[status]:
		out.add(4)
	case armEnded[status] != hasEnd:
		out.add(2)
	}
	if armTimestampOK(run["start"]) && armTimestampOK(run["end"]) {
		if run["end"].(string) < run["start"].(string) {
			out.add(3)
		}
	}
	if steps, ok := run["steps"].([]any); ok {
		if count, isInt := intValue(run["step_count"]); isInt && count < int64(len(steps)) {
			out.add(5)
		}
	}
}

func armRowsRunMembers(run map[string]any, out armRejects) {
	root, hasRoot := run["root_run_id"]
	_, hasParent := run["parent_run_id"]
	if hasRoot && !hasParent && !jsonEqual(root, run["run_id"]) {
		out.add(16)
	}
	for member := range run {
		if !armRunMembers[member] && !strings.HasPrefix(member, "x-") {
			out.add(21)
			break
		}
	}
	if raw, present := run["labels"]; present && raw != nil {
		labels, ok := raw.(map[string]any)
		if !ok {
			out.add(18)
		} else {
			for _, v := range labels {
				if !isStr(v) {
					out.add(18)
					break
				}
			}
		}
	}
	if cost, ok := run["cost"].(map[string]any); ok {
		amount, isString := cost["amount"].(string)
		if !isString || !armAmount.MatchString(amount) {
			out.add(17)
		}
	}
}

func armRowsTimestampsAndIdentifiers(run map[string]any, out armRejects) {
	stamps := []any{run["start"], run["end"]}
	for _, step := range armSteps(run) {
		stamps = append(stamps, step["start"], step["end"])
	}
	for _, s := range stamps {
		if s != nil && !armTimestampOK(s) {
			out.add(19)
			break
		}
	}
	var names []any
	for _, key := range armIdentifiers {
		if v, present := run[key]; present {
			names = append(names, v)
		}
	}
	for _, step := range armSteps(run) {
		if v, present := step["invocation_id"]; present {
			names = append(names, v)
		}
		if tool, ok := step["tool"].(map[string]any); ok {
			if v, present := tool["name"]; present {
				names = append(names, v)
			}
		}
	}
	for _, n := range names {
		s, ok := n.(string)
		if !ok || len([]rune(s)) == 0 || len([]rune(s)) > 128 {
			out.add(20)
			break
		}
	}
}

// armCounters reads a Usage object's counters as integers, reporting whether
// every one is a non-negative integer.
func armCounters(usage map[string]any) (map[string]int64, bool) {
	counters := map[string]int64{}
	for key, v := range usage {
		if key == "cache_writes" {
			continue
		}
		n, ok := intValue(v)
		if !ok || n < 0 {
			return nil, false
		}
		counters[key] = n
	}
	return counters, true
}

func armRowsUsage(value any, out armRejects) {
	usage, ok := value.(map[string]any)
	if !ok {
		return
	}
	counters, ok := armCounters(usage)
	if !ok {
		out.add(9)
		return
	}
	read, write := counters["cache_read_tokens"], counters["cache_write_tokens"]
	if input, present := counters["input_tokens"]; present {
		if read > input {
			out.add(10)
		} else if read+write > input {
			out.add(11)
		}
	}
	if output, present := counters["output_tokens"]; present && counters["reasoning_tokens"] > output {
		out.add(22)
	}
	writes, ok := usage["cache_writes"].([]any)
	if !ok {
		return
	}
	lifetimes := map[string]bool{}
	duplicate := false
	var sum int64
	allInts := true
	for _, raw := range writes {
		item, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		lifetime, _ := item["lifetime"].(string)
		if lifetimes[lifetime] {
			duplicate = true
		}
		lifetimes[lifetime] = true
		tokens, isInt := intValue(item["tokens"])
		if !isInt {
			allInts = false
		}
		sum += tokens
	}
	if duplicate {
		out.add(13)
	}
	if allInts && sum != write {
		out.add(14)
	}
}

func armRowsSteps(run map[string]any, out armRejects) {
	steps := armSteps(run)
	if !armStepsOrdered(steps) {
		out.add(6)
	}
	invocations := map[string]bool{}
	repeated := false
	for _, step := range steps {
		armRowsStepKind(step, out)
		if id, ok := step["invocation_id"].(string); ok && step["kind"] == "model_invocation" {
			if invocations[id] {
				repeated = true
			}
			invocations[id] = true
		}
		armRowsUsage(step["usage"], out)
	}
	if repeated {
		out.add(15)
	}
}

// armStepsOrdered: indexes strictly increase and consecutive starts are both
// strings and non-decreasing, as the Python rail reads them; a lone step with
// no start is not out of order with anything.
func armStepsOrdered(steps []map[string]any) bool {
	seen := map[int64]bool{}
	ordered := true
	var previous int64
	var previousStart string
	previousOK := false
	for i, step := range steps {
		index, _ := intValue(step["index"])
		start, startOK := step["start"].(string)
		if seen[index] || (i > 0 && index < previous) {
			ordered = false
		}
		if i > 0 && (!previousOK || !startOK || previousStart > start) {
			ordered = false
		}
		seen[index] = true
		previous, previousStart, previousOK = index, start, startOK
	}
	return ordered
}

// armRowsStepKind: a model invocation carries usage and a model object; a
// tool call carries a tool object and no usage.
func armRowsStepKind(step map[string]any, out armRejects) {
	kind, _ := step["kind"].(string)
	switch kind {
	case "model_invocation":
		if !isObj(step["usage"]) || !isObj(step["model"]) {
			out.add(7)
		}
	case "tool_call":
		_, hasUsage := step["usage"]
		if !isObj(step["tool"]) || hasUsage {
			out.add(8)
		}
	}
}

func armRowsTotals(run map[string]any, out armRejects) {
	totals, totalsOK := run["totals"].(map[string]any)
	armRowsUsage(run["totals"], out)
	steps, ok := run["steps"].([]any)
	count, countOK := intValue(run["step_count"])
	if !ok || !totalsOK || !countOK || count != int64(len(steps)) {
		return
	}
	sums := map[string]int64{}
	members := map[string]bool{}
	for _, step := range armSteps(run) {
		usage, ok := step["usage"].(map[string]any)
		if !ok {
			continue
		}
		counters, ok := armCounters(usage)
		if !ok {
			return
		}
		for key, n := range counters {
			sums[key] += n
			members[key] = true
		}
	}
	for key := range totals {
		if key != "cache_writes" {
			members[key] = true
		}
	}
	keys := make([]string, 0, len(members))
	for key := range members {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, key := range keys {
		if !numberEquals64(totals[key], sums[key]) {
			out.add(12)
			return
		}
	}
}

// numberEquals64 is numberEquals over an int64 expectation; a missing member
// is never equal, as Python's totals.get(member) != expected reads it.
func numberEquals64(v any, want int64) bool {
	number, ok := v.(json.Number)
	if !ok {
		return false
	}
	f, err := number.Float64()
	return err == nil && f == float64(want)
}

func armRejections(document map[string]any) []string {
	out := armRejects{}
	armRowsRunStatus(document, out)
	armRowsRunMembers(document, out)
	armRowsTimestampsAndIdentifiers(document, out)
	armRowsSteps(document, out)
	armRowsTotals(document, out)
	ids := make([]string, 0, len(out))
	for id := range out {
		ids = append(ids, id)
	}
	sort.Strings(ids)
	return ids
}
